feelpp/feel/feelcrb/pbdw.hpp

```diff
@@ -0,0 +1,355 @@
+        this->setDBDirectory(Environment::randomUUID(true));
+
+    if( ! M_rebuildDb )
+        this->loadDB(this->absoluteDbFilename(), crb::load::all );
+}
+
+template<typename RBSpace>
+void
+PBDW<RBSpace>::offline()
+{
+    Feel::cout << "offline phase start with M=" << M_M << " and N=" << M_N << std::endl;
+    auto u = M_XR->functionSpace()->element();
+    std::vector<element_type> qs(M_M, M_XR->functionSpace()->element());
+    auto a = form2(_test=M_XR->functionSpace(), _trial=M_XR->functionSpace());
+    a = integrate(_range=elements(M_XR->mesh()), _expr=inner(vf::id(u),vf::idt(u)) + inner(grad(u),gradt(u)) );
+    auto am = a.matrixPtr();
+    for( int m = 0; m < M_M; ++m )
+    {
+        auto f = form1(_test=M_XR->functionSpace(), _vector=M_sigmas[m]->containerPtr());
+        a.solve(_solution=qs[m], _rhs=f, _name="pbdw");
+    }
+    M_matrix = matrixN_type::Zero(M_M+M_N, M_M+M_N);
+    for(int i = 0; i < M_M; ++i )
+    {
+        for(int j = 0; j < i; ++j )
+        {
+            M_matrix(i, j) = am->energy(qs[i], qs[j]);
+            M_matrix(j, i) = M_matrix(i, j);
+        }
+        M_matrix(i, i) = am->energy(qs[i], qs[i]);
+        for(int j = 0; j < M_N; ++j )
+            M_matrix(i, M_M+j) = am->energy(qs[i], M_XR->primalBasisElement(j));
+    }
+    M_matrix.bottomLeftCorner(M_N, M_M) = M_matrix.topRightCorner(M_M, M_N).transpose();
+
+    this->saveDB();
+}
+
+template<typename RBSpace>
+typename PBDW<RBSpace>::vectorN_type
+PBDW<RBSpace>::online(vectorN_type const& yobs) const
+{
+    vectorN_type yobs2 = vectorN_type::Zero(this->dimension());
+    yobs2.head(this->dimensionM()) = yobs;
+    vectorN_type vn = M_matrix.colPivHouseholderQr().solve(yobs2);
+    return vn;
+}
+
+template<typename RBSpace>
+typename PBDW<RBSpace>::element_type
+PBDW<RBSpace>::solution(vectorN_type const& yobs) const
+{
+    int M = this->dimensionM();
+    int N = this->dimensionN();
+    auto vn = this->online(yobs);
+    auto wn = M_XR->primalRB();
+    auto Xh = M_XR->functionSpace();
+    auto I = Xh->element();
+    I.zero();
+    for(int i = 0; i < M; ++i )
+        I.add(vn(i), Xh->element(M_sigmas[i]->containerPtr()));
+    for(int i = 0; i < N; ++i )
+        I.add(vn(M+i), unwrap_ptr(wn[i]));
+    return I;
+}
+
+template<typename RBSpace>
+void
+PBDW<RBSpace>::saveDB()
+{
+    if ( this->worldComm().isMasterRank() )
+    {
+        if( fs::exists(this->dbLocalPath()) )
+        {
+            if( !fs::is_directory(this->dbLocalPath()) )
+                throw std::exception();
+        }
+        else
+            fs::create_directories(this->dbLocalPath());
+
+        json j;
+        j["name"] = this->name();
+        std::time_t t = std::time(nullptr);
+        std::stringstream ss;
+        ss << std::put_time(std::localtime(&t), "%c %Z");
+        j["date"] = ss.str();
+        j["mesh"] = this->absoluteMeshFilename();
+        std::ofstream o(this->absoluteJsonFilename());
+        o << j << std::endl;
```

Comment:
use 
```
j.dump(1) 
```
which is better for reading

Reply:
or maybe is it already nicely indented ?

Reply:
no it is all on one line. I'll change it.